File: 백준/Silver/1463. 1로 만들기/1로 만들기.py
import sys
input = sys.stdin.readline

n = int(input())

D = [0] * (n+1)

D[1] = 0    # 0으로 초기화 해서 필요없긴 함...

for i in range(2, n+1):
    D[i] = D[i-1] + 1
    if i % 2 == 0:
        D[i] = min(D[i], D[i//2] + 1)
    if i % 3 == 0:
        D[i] = min(D[i], D[i//3] + 1)
        
print(D[n])

File: 백준/Silver/1463. 1로 만들기/1로 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	static Integer[] dp;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		
		dp = new Integer[n+1];
		dp[0] = dp[1] = 0;
		System.out.println(cal(n));
	}
	
	static int cal(int n) {
		if(dp[n] == null) {
			if(n % 6 == 0)
				dp[n] = Math.min(cal(n/3), Math.min(cal(n/2), cal(n-1))) + 1;
			else if(n % 3 == 0)
				dp[n] = Math.min(cal(n/3), cal(n-1)) + 1;
			else if(n % 2 == 0) 
				dp[n] = Math.min(cal(n/2), cal(n-1)) + 1;
			else
				dp[n] = cal(n-1) + 1;
		}
		return dp[n];
	}
}